[c10d][Sym mem] Add set_signal_pad_size API for SymmetricMemory

yang-yu-hang opened the pull request:
Summary:
The signal pad size for symmetric memory was previously hardcoded as a constexpr, which may be too small for workloads that launch a large number of blocks. This change exposes `set_signal_pad_size` and `get_signal_pad_size` APIs to allow users to configure the signal pad size before making symmetric memory allocations.

### Changes:

#### 1. Core API (C++)
- Renamed `signal_pad_size` constexpr to `default_signal_pad_size` in CUDASymmetricMemoryTypes.hpp
- Added `get_signal_pad_size()` and `set_signal_pad_size(size_t)` function declarations in CUDASymmetricMemoryTypes.hpp
- Implemented the functions in SymmetricMemory.cpp using `std::optional<size_t>` to distinguish between default and user-configured values
- Added TORCH_API exports in SymmetricMemory.hpp for public API access

#### 2. Backend Updates
- Updated CUDASymmetricMemory.cu to call `get_signal_pad_size()` instead of using the hardcoded constant
- Updated NCCLSymmetricMemory.cu to use configurable signal pad size with local variable `signal_pad_size`
- Updated NVSHMEMSymmetricMemory.cu to use configurable signal pad size with local variable `signal_pad_size`

#### 3. Python Bindings
- Added Python bindings in init.cpp with comprehensive docstrings explaining usage
- Added Python wrapper functions in torch/distributed/_symmetric_memory/__init__.py
- Updated `__all__` to export the new API functions

#### 4. Tests
- Added `test_get_signal_pad_size()` to verify the API returns a positive integer and Python/C++ consistency
- Added `test_set_signal_pad_size()` to verify setting, getting, and restoring signal pad size values


Test Plan:
`PYTHONPATH=. python3 test/distributed/test_symmetric_memory.py`

ngimel reviewed (commented):
Looks good, a few minor comments. Thanks!

yang-yu-hang commented on torch/csrc/distributed/c10d/init.cpp:
```diff
@@ -1137,6 +1137,12 @@ This class does not support ``__members__`` property.)");
           &::c10d::symmetric_memory::has_multicast_support)
       .def_static("set_backend", &::c10d::symmetric_memory::set_backend)
       .def_static("get_backend", &::c10d::symmetric_memory::get_backend)
+      .def_static(
```
Good point! I have changed `def_property_readonly` to `def_property_static` for "signal_pad_size".

cyyever commented:
@pytorchbot rebase -b main

yang-yu-hang commented:
@pytorchbot rebase -b main

yang-yu-hang commented:
@pytorchmergebot merge

yang-yu-hang committed:
[c10d][Sym mem] Add set_signal_pad_size API for SymmetricMemory

Summary:
The signal pad size for symmetric memory was previously hardcoded as a constexpr, which may be too small for workloads that launch a large number of blocks. This change exposes `set_signal_pad_size` and `get_signal_pad_size` APIs to allow users to configure the signal pad size before making symmetric memory allocations.

### Changes:

#### 1. Core API (C++)
- Renamed `signal_pad_size` constexpr to `default_signal_pad_size` in CUDASymmetricMemoryTypes.hpp
- Added `get_signal_pad_size()` and `set_signal_pad_size(size_t)` function declarations in CUDASymmetricMemoryTypes.hpp
- Implemented the functions in SymmetricMemory.cpp using `std::optional<size_t>` to distinguish between default and user-configured values
- Added TORCH_API exports in SymmetricMemory.hpp for public API access

#### 2. Backend Updates
- Updated CUDASymmetricMemory.cu to call `get_signal_pad_size()` instead of using the hardcoded constant
- Updated NCCLSymmetricMemory.cu to use configurable signal pad size with local variable `signal_pad_size`
- Updated NVSHMEMSymmetricMemory.cu to use configurable signal pad size with local variable `signal_pad_size`

#### 3. Python Bindings
- Added Python bindings in init.cpp with comprehensive docstrings explaining usage
- Added Python wrapper functions in torch/distributed/_symmetric_memory/__init__.py
- Updated `__all__` to export the new API functions

#### 4. Tests
- Added `test_get_signal_pad_size()` to verify the API returns a positive integer and Python/C++ consistency
- Added `test_set_signal_pad_size()` to verify setting, getting, and restoring signal pad size values


Test Plan:
Build and existing symmetric memory tests. The API follows the same pattern as existing `set_backend`/`get_backend` APIs.

yang-yu-hang commented:
@pytorchmergebot merge
